core/local/local_test.go

```diff
@@ -974,6 +974,98 @@ func TestExecutionSchedulerIsRunning(t *testing.T) {
 	assert.NoError(t, <-err)
 }
 
+type mockDNSRunner struct {
+	*js.Runner
+	resolver *testutils.MockResolver
+}
+
+// NewVU initializes a new VU replacing its resolver with a MockResolver.
+func (r *mockDNSRunner) NewVU(id int64, out chan<- stats.SampleContainer) (lib.InitializedVU, error) {
+	initVU, err := r.Runner.NewVU(id, out)
+	if err != nil {
+		return nil, err
+	}
+	(initVU.(*js.VU)).Dialer.Resolver = r.resolver
+	return initVU, nil
+}
+
+func TestDNS(t *testing.T) {
+	t.Parallel()
+	tb := httpmultibin.NewHTTPMultiBin(t)
+	defer tb.Cleanup()
+	sr := tb.Replacer.Replace
+	script := sr(`
+		import http from "k6/http";
+		import { sleep } from "k6";
+
+		export let options = {
+			vus: 1,
+			duration: "5s",
+			noConnectionReuse: true,
+		}
+
+		export default function () {
+			const res = http.get("http://myhost:HTTPBIN_PORT/", { timeout: 50 });
+			sleep(0.45);  // not an even multiple of 0.5 to minimize races with asserts
+		}`)
+
+	t.Run("cache", func(t *testing.T) {
+		testCases := map[string]struct {
+			opts lib.Options
+		}{
+			"inf": { // IPs are cached indefinitely
+				lib.Options{},
+			},
+			"no": { // every request does a DNS lookup
+				lib.Options{DNS: &lib.DNSConfig{
+					TTL:      null.StringFrom("0"),
+					Strategy: lib.DefaultDNSConfig.Strategy,
+				}},
+			},
+			"2s": {
+				lib.Options{DNS: &lib.DNSConfig{
+					TTL:      null.StringFrom("2s"),
+					Strategy: lib.DefaultDNSConfig.Strategy,
+				}},
+			},
+		}
+
+		for name, tc := range testCases {
+			tc := tc
+			t.Run(name, func(t *testing.T) {
+				logger := logrus.New()
+				logger.SetOutput(testutils.NewTestOutput(t))
+				runner, err := js.New(logger, &loader.SourceData{
+					URL: &url.URL{Path: "/script.js"}, Data: []byte(script),
+				}, nil, lib.RuntimeOptions{})
+				require.NoError(t, err)
+
+				mr := &mockDNSRunner{runner, testutils.NewMockResolver(nil)}
+				ctx, cancel, execScheduler, samples := newTestExecutionScheduler(t, mr, logger, tc.opts)
+				defer cancel()
+
+				mr.resolver.Set("myhost", sr("HTTPBIN_IP"))
+				time.AfterFunc(1*time.Second, func() {
+					mr.resolver.Set("myhost", "127.0.0.254")
+				})
+				errCh := make(chan error, 1)
+				go func() { errCh <- execScheduler.Run(ctx, ctx, samples) }()
+
+				for {
+					select {
+					case err := <-errCh:
+						require.NoError(t, err)
+						return
+					case <-time.After(6 * time.Second):
+						t.Fatal("timed out")
+						return
+					}
+				}
```

Comment:
Yeah, as I mentioned in the description, these tests are half-baked / useless right now without properly mocking the DNS server. The idea of this and the other `TestVUIntegrationDNS` test was to start the execution scheduler with a valid IP of the httpmultibin server, then update the IP to a failing one and assert on what the behavior should be with the different TTL/strategy options. But since I wasn't able to mock it properly, I left them here purely for reference.

But thanks for mentioning mocking of `Resolver.Dial`. If we bring in dnscache internally then this test would be possible, so I'll see what I can do.

Reply:
The test is fixed in the latest changes, though I still want to get rid of the global var mocking, and the timings might be flaky, but I don't see a way around that.